org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/MavenBuildSupport.java

```diff
@@ -25,24 +39,84 @@
  *
  */
 public class MavenBuildSupport implements IBuildSupport {
+	private Map<String, String> pomDigests;
+
+	private static final String POM_SERLIZATION_FILE_NAME = ".pom-digests";
+
+	private IProjectConfigurationManager configurationManager;
+
+	public MavenBuildSupport(IProjectConfigurationManager configurationManager) {
+		this.configurationManager = configurationManager;
+		File f = getStateFile();
+		if (f.isFile()) {
+			pomDigests = deserializePomDigests(f);
+		} else {
+			pomDigests = new HashMap<>();
+		}
+
+	}
+
+	/**
+	 * @return
+	 */
+	private File getStateFile() {
+		return JavaLanguageServerPlugin.getInstance().getStateLocation().append(POM_SERLIZATION_FILE_NAME).toFile();
+	}
 
 	@Override
 	public boolean applies(IProject project) {
 		return ProjectUtils.isMavenProject(project);
 	}
 
 	@Override
-	public void update(IProject project, IProgressMonitor monitor) throws CoreException {
-		if (!applies(project)) {
+	public void update(IProject project, boolean force, IProgressMonitor monitor) throws CoreException {
+		if (!applies(project) || (!needsMavenUpdate(project) && !force)) {
 			return;
 		}
-		JavaLanguageServerPlugin.logInfo("Starting Maven update for "+project.getName());
+		JavaLanguageServerPlugin.logInfo("Starting Maven update for " + project.getName());
 		//TODO collect dependent projects and update them as well? i.e in case a parent project was modified
-		IProjectConfigurationManager configurationManager = MavenPlugin.getProjectConfigurationManager();
 		MavenUpdateRequest request = new MavenUpdateRequest(project, MavenPlugin.getMavenConfiguration().isOffline(), true);
 		configurationManager.updateProjectConfiguration(request, monitor);
 	}
 
+	private boolean needsMavenUpdate(IProject project) {
```

Comment:
the more I think of it the less I like this implementation. This method should simply check if a project needs an update. Updating the  digests map should be done once the update is effectively triggered

Reply:
But the two are really equivalent here, no? If the the digest is changed and an update needed, the project WILL be updated. So while not pretty, I think the logic is correct.

Reply:
it works for now. Until needsMavenUpdate is called from somewhere else (even in the same class since it's private) and no update is performed, then things will go wrong. 
calling needsMavenUpdate twice in a row should yield the same results, always. Else, please change the method name to indicate it does more than checking.